Code/CryMP/Client/WeatherSystem.cpp

```diff
@@ -0,0 +1,438 @@
+#include "WeatherSystem.h"
+#include "CryCommon/CrySystem/ISystem.h"
+#include "CryCommon/CrySystem/ICryPak.h"
+#include "CryCommon/CrySystem/IConsole.h"
+#include "CryCommon/Cry3DEngine/IMaterial.h"
+#include "CryCommon/CryEntitySystem/IEntitySystem.h"
+#include "CryCommon/CryAction/IVehicleSystem.h"
+#include "CryCommon/CryAction/IGameTokens.h"
+#include "CryCommon/CryMath/Cry_Camera.h"
+#include "CryCommon/CrySystem/ILog.h"
+#include "CrySystem/LocalizationManager.h"
+#include "CrySystem/RandomGenerator.h"
+#include "Library/StringTools.h"
+#include "Library/Util.h"
+#include "Library/WinAPI.h"
+
+#include "ScriptBind_CPPAPI.h"
+#include "Client.h"
+#include "ScriptCommands.h"
+#include "ScriptCallbacks.h"
+#include "DrawTools.h"
+#include "CryGame/GameActions.h"
+#include "CryGame/Actors/Actor.h"
+
+#include <map>
+
+static bool						currently_enabled = false;
+static std::map<int, float[3]>	original_weather_values;
+static std::optional<Vec3>		original_wind;
+
+static std::map<int, string>	active_values;
+static std::set<std::string>	active_effects;
+static std::optional<unsigned>  active_mask;
+
+static constexpr std::array<EERType, 3> static_entities = {
+	eERType_Vegetation,
+	eERType_Brush,
+	eERType_Rope
+};
+
+static const int WEATHER_NAMESPACE		= 2000;
+static const int WEATHER_ENV_NAMESPACE  = 2100;
+
+enum EWeatherVars {
+	WEATHER_ENABLED,
+	WEATHER_WIND,
+	WEATHER_FX,
+	WEATHER_LAYERS,
+	WEATHER_VAR_COUNT
+};
+
+CWeatherSystem::CWeatherSystem() {
+	original_weather_values.clear();
+	m_time = 0;
+	m_lastUpdate = -1000.0f;
+}
+
+void CWeatherSystem::Reset() {
+	RestoreWeather(true);
+	if (original_wind) {
+		gEnv->p3DEngine->SetWind(*original_wind);
+	}
+
+	for (auto& k : active_effects) {
+		if (k[0] == '-') SpawnEffect(k.c_str() + 1);
+		else RemoveEffect(k.c_str());
+	}
+
+	if (active_mask) {
+		RemoveLayer(*active_mask);
+		active_mask.reset();
+	}
+	
+	currently_enabled = false;
+	original_weather_values.clear();
+	active_values.clear();
+	original_wind.reset();
+	active_effects.clear();
+}
+
+void CWeatherSystem::Update(float frameTime) {
+	CServerSynchedStorage *pSSS = g_pGame->GetServerSynchedStorage();
+	if (!pSSS) {
+		return;
+	}
+
+	m_time += frameTime;
+
+	string b;
+	if (!pSSS->GetGlobalValue(WEATHER_NAMESPACE + WEATHER_ENABLED, b) || b.length() == 0 || b == "0") {
```

Comment:
```suggestion
	bool enabled = false;
        pSS->GetGlobalValue(WEATHER_NAMESPACE + WEATHER_ENABLED, enabled);
        if (!enabled) {
```

Reply:
Updated this one, with CSynchedStorage it works, but I would still leave those strings in place instead of float as it allows for greater flexibility. Some TOD fields are floats, some are vectors, also having option to use either raw value `r123.45` ot interpolated `123.45` makes things much easier in case you don't remember correct `min`, `max` values for each TOD 😄

Reply:
Sounds good 👍 
GetObjectsByType(eERType_Vegetation) returns several thousand rendernodes for me, on mesa its 100k :D it doesn't for you? 
You can specify eERType_RoadObject_NEW (not the other one) and it will return roads as well

Reply:
The athmosphere is really cool, with the snow effects. Especially at night :)

Reply:
Well, it works now, all of sudden, I guess I just had old code (created this branch month ago, didn't sync with master :D), but it's weird, still with MTL_LAYER_FROZEN vegetation isn't frozen. I even tried using DYNFROZEN, but it didn't help.
Gotta find out how it works in Sandbox editor :D As there it clearly works when you enable Layer_frozen for vegetation

Reply:
I made entities it should freeze / wet an global array (static_entities) so feel free to experiment with that on what everything to freeze, as of not there is only three types (brushes, vegeatation, ropes) 😄

Reply:
Yea. Just dumped mesa stats. Actually its almost 300k vegetations on Mesa.. :D
Freezing everything would be super cool literally ;)

EERType: Unknown, Count: 685
EERType: Brush, Count: 6729
EERType: Vegetation, Count: 290768
EERType: Light, Count: 66
EERType: Cloud, Count: 118
EERType: FogVolume, Count: 3
EERType: Decal, Count: 1060
EERType: ParticleEmitter, Count: 77
EERType: WaterVolume, Count: 2
EERType: RoadObject_NEW, Count: 944
EERType: DistanceCloud, Count: 119

Reply:
I guess I figured it out! :D
![ScreenShot0048](https://github.com/user-attachments/assets/29da4a9b-42ff-4a60-9b2a-d37e18fd2fb6)

Just gotta fix some issues to make it more pretty :D

Reply:
Omg 👯

Reply:
Pushed it, feel free to try out. Trick is that vegetation is actually an instanced stat object, so this code just naively iterates all stat instance group IDs from 0 to 65536 (sadly there is no GetGroupIDs, but let's assume it's never like 1 000 000+) and sets material flags on those instead.